deform/linear_deformation.py

```diff
@@ -0,0 +1,1173 @@
+
+    def __init__(self, template, displacement):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        displacement: `ProductSpaceVector`
+            Linearized deformation parameters for image grid points.
+        template : `DiscreteLpVector`
+            Fixed template deformed by the vector field. Its space
+            must have the same number of dimensions as ``par_space``.
+        """
+
+        super().__init__(template)
+
+        self.template = template
+        self.displacement = displacement
+
+        Operator.__init__(self, self.displacement.space, self.template.space,
+                          linear=True)
+
+    def derivative(self, *args, **kwargs):
+        raise NotImplementedError
+
+    def _deform_grad(self, grad_f):
+        """Compute the deformation of template's gradient.
+
+        deformed template's gradient: \gradient I(x + v(x))
+
+        Parameters
+        ----------
+        grad_f: `ProductSpaceVector`
+            Gradient of the template, i.e. a vector field on the
+            image domain
+        """
+        temp = [self.linear_deform(gf, self.displacement) for gf in grad_f]
+        return self.displacement.space.element(temp)
+
+    def _call(self, disp_like):
+        """Implementation of ``self(disp_like)``.
+
+        Parameters
+        ----------
+        disp_like: `ProductSpaceVector`
+            Like the linearized deformation parameters for image grid points.
+        """
+        grad = odl.Gradient(self._range)
+        template_grad = grad(self.template)
+        def_grad = self._deform_grad(template_grad)
+
+        result = def_grad[0] * disp_like[0].reshape(self.image_grid.shape,
+                                                    order=self.range.order)
+        for gf, d in zip(def_grad[1:], disp_like[1:]):
+            result += gf * d.reshape(self.image_grid.shape,
+                                     order=self.range.order)
+        return result
+
+    @property
+    def adjoint(self):
+        """Adjoint of the template deformation derivative."""
+        adj_op = LinearizedDeformationDerivativeAdjoint(
+            self.template, self.displacement)
+        return adj_op
+
+
+class LinearizedDeformationDerivativeAdjoint(LinearizedDeformationDerivative):
+
+    """Adjoint of the template deformation operator derivative.
+
+    TODO: write a bit more
+    """
+    def __init__(self, template, displacement):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        template : `DiscreteLpVector`
+            Fixed template deformed by the vector field. Its space
+            must have the same number of dimensions as ``par_space``.
+        displacement: `ProductSpaceVector`
+            Linearized deformation parameters for image grid points.
+        """
+        super().__init__(template, displacement)
+
+        # Switch domain and range
+        self._domain, self._range = self._range, self._domain
```

Comment:
OK, I see what you mean. Basically the operator cannot consider `domain` and `range` as "its own" but rather something that was given from the parent class(es). In any case, we should consider the "common ancestor" principle then, seem to be the best (most explicit) way to avoid tons of duplicate code.